monitoring/queuemon.go

```diff
@@ -0,0 +1,204 @@
+	done     chan struct{}
+	// handler for the event queue
+	queue outqueue.Queue
+	log   *logp.Logger
+	// A awkward no-op if a user has disabled monitoring
+	bypass bool
+
+	// Count of times the queue has reached a configured limit.
+	queueLimitCount uint64
+}
+
+//Config is the intermediate struct representation of the queue monitor config
+type Config struct {
+	Outputs  OutputConfig  `config:"outputs"`
+	Interval time.Duration `config:"interval"`
+	Enabled  bool          `config:"enabled"`
+}
+
+// OutputConfig is the main config subsection for the
+type OutputConfig struct {
+	LogOutput    log.Config           `config:"log"`
+	ExpvarOutput expvar.ExpvarsConfig `config:"expvar"`
+}
+
+// DefaultConfig returns the default settings for the queue monitor
+func DefaultConfig() Config {
+	return Config{
+		Outputs: OutputConfig{
+			LogOutput: log.Config{
+				Enabled: true,
+			},
+			ExpvarOutput: expvar.ExpvarsConfig{
+				Enabled: false,
+				Addr:    ":8080",
+				Name:    "queue",
+			},
+		},
+
+		Enabled:  true,
+		Interval: time.Second * 30,
+	}
+}
+
+// NewFromConfig creates a new queue monitor from a pre-filled config struct.
+func NewFromConfig(cfg Config, queue outqueue.Queue) (*QueueMonitor, error) {
+	// the queue == nil is largely a shim to make things not panic while we wait for the queues to get hooked up.
+	if !cfg.Enabled || queue == nil {
+		return &QueueMonitor{bypass: true}, nil
+	}
+	//init outputs
+	outputs := initOutputs(cfg)
+	return &QueueMonitor{
+		interval: cfg.Interval,
+		queue:    queue,
+		done:     make(chan struct{}),
+		log:      logp.L(),
+		outputs:  outputs,
+	}, nil
+}
+
+// Watch is a non-blocking call that starts up a queue watcher that will report metrics to a given output
+func (mon QueueMonitor) Watch() {
+	// Turn this function into a no-op if nothing is initialized.
+	if mon.bypass {
+		return
+	}
+	ticker := time.NewTicker(mon.interval)
+	go func() {
+		for {
+			select {
+			case <-mon.done:
+				return
+			case <-ticker.C:
+				//We're assuming that the `Metrics()` call from the queue won't hard-block.
+				err := mon.updateMetrics()
+				if err != nil {
+					mon.log.Errorf("Error updating metrics: %w", err)
+				}
+			}
+		}
+	}()
+}
+
+// End closes the watcher
+// As of now this isn't being called outside of tests, as don't have any kind of signal catch/shutdown in the shipper itself.
+func (mon QueueMonitor) End() {
+	if mon.bypass {
+		return
+	}
+	mon.done <- struct{}{}
+}
+
+// updateMetrics is responsible for fetching the metrics from the queue, calculating whatever it needs to, and sending the complete events to the output
+func (mon *QueueMonitor) updateMetrics() error {
+	raw, err := mon.queue.Metrics()
+	if err != nil {
+		return fmt.Errorf("error fetching queue Metrics: %w", err)
+	}
+
+	count, limit, queueIsFull, err := getLimits(raw)
+	if err != nil {
+		return fmt.Errorf("could not get limits: %w", err)
+	}
+
+	if queueIsFull {
+		mon.queueLimitCount = mon.queueLimitCount + 1
```

Comment:
Reading this, what information is this providing us with that `QueueIsCurrentlyFull` doesn't give us?

I think where we would this is in the shipper Publish RPC handler, so we could track how many incoming events were rejected because the queue is full. I'm not sure how useful it is here.